RecoLocalCalo/HcalRecAlgos/src/DoMahiAlgo.cc

```diff
@@ -0,0 +1,563 @@
+  BXSizeConf_   = activeBXs_.size();
+}
+
+void DoMahiAlgo::phase1Apply(const HBHEChannelInfo& channelData,
+			     float& reconstructedEnergy,
+			     float& reconstructedTime,
+			     float& chi2) {
+
+  TSSize_ = channelData.nSamples();
+  TSOffset_ = channelData.soi();
+  FullTSOffset_ = FullTSofInterest_ - TSOffset_;
+
+  // 1 sigma time constraint
+  if (channelData.hasTimeInfo()) dt_=timeSigmaSiPM_;
+  else dt_=timeSigmaHPD_;
+
+  //fC to photo-electron scale factor (for P.E. uncertainties)
+  fcByPe_ = channelData.fcByPE();
+
+  //Dark current value for this channel (SiPM only)
+  darkCurrent_ =  getSiPMDarkCurrent(channelData.darkCurrent(), 
+				     channelData.fcByPE(),
+				     channelData.lambda());
+
+  //Average pedestal width (for covariance matrix constraint)
+  pedConstraint_ = 0.25*( channelData.tsPedestalWidth(0)*channelData.tsPedestalWidth(0)+
+			  channelData.tsPedestalWidth(1)*channelData.tsPedestalWidth(1)+
+			  channelData.tsPedestalWidth(2)*channelData.tsPedestalWidth(2)+
+			  channelData.tsPedestalWidth(3)*channelData.tsPedestalWidth(3) );
+
+  std::vector<float> reconstructedVals;
+  SampleVector charges;
+  
+  double tsTOT = 0, tstrig = 0; // in GeV
+  for(unsigned int iTS=0; iTS<TSSize_; ++iTS){
+    double charge = channelData.tsRawCharge(iTS);
+    double ped = channelData.tsPedestal(iTS);
+
+    charges.coeffRef(iTS) = charge - ped;
+
+    //ADC granularity
+    double noiseADC = (1./sqrt(12))*channelData.tsDFcPerADC(iTS);
+
+    //Dark current (for SiPMs)
+    double noiseDC=0;
+    if(channelData.hasTimeInfo() && !channelData.hasEffectivePedestals() && (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noiseDC = darkCurrent_;
+    }
+
+    //Photostatistics
+    double noisePhoto = 0;
+    if ( (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noisePhoto = sqrt((charge-ped)*channelData.fcByPE());
+    }
+
+    //Electronic pedestal
+    double pedWidth = channelData.tsPedestalWidth(iTS);
+
+    //Total uncertainty from all sources
+    noiseTerms_.coeffRef(iTS) = noiseADC*noiseADC + noiseDC*noiseDC + noisePhoto*noisePhoto + pedWidth*pedWidth;
+
+    tsTOT += charge - ped;
+    if( iTS==TSOffset_ ){
+      tstrig += (charge - ped)*channelData.tsGain(0);
+    }
+  }
+
+  bool status =false;
+  if(tstrig >= TS4Thresh_) {
+    if (chiSqSwitch_>0) {
+      status = DoFit(charges,reconstructedVals,1);
+      if (reconstructedVals[1]>chiSqSwitch_) {
+	status = DoFit(charges,reconstructedVals,0);
+      }
+    }
+    else {
+      status = DoFit(charges,reconstructedVals,0);
+    }
+  }
+  
+  if (!status) {
+    reconstructedVals.clear();
+    reconstructedVals.push_back(0.);
+    reconstructedVals.push_back(888.);
+  }
+  
+  reconstructedEnergy = reconstructedVals[0]*channelData.tsGain(0);
+  reconstructedTime = 0;
+  chi2 = reconstructedVals[1];
+
+}
+
+bool DoMahiAlgo::DoFit(SampleVector amplitudes, std::vector<float> &correctedOutput, int nbx) {
```

Comment:
can you pass `amplitudes` by `const&` ?
or, if it has an efficient move assignment operator, change line 156 to
```c++
amplitudes_ = std::move(amplitudes);
```
?